server/src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -1447,6 +1447,8 @@ protected void processResponse(TransportResponseHandler handler, TransportRespon
                 handler.handleResponse(response);
             } catch (Exception e) {
                 processException(handler, wrapInRemote(new ResponseHandlerFailureTransportException(e)));
+            } finally {
+                response.decRef();
```

Comment:
I think we should also release this ref on the other paths through `sendResponse`:

- `shutdownBlock == null`
- `handler == null`
- `ForkingResponseHandlerRunnable` is rejected

Reply:
Right, I borrowed some of your suggested code. 

I added a test for the bug that interests me (ProxyResponseHandler with `SAME` thread pool). Other paths are not tested.